superset/common/query_context_processor.py

```diff
@@ -685,22 +689,53 @@ def get_native_annotation_data(query_obj: QueryObject) -> dict[str, Any]:
     def get_viz_annotation_data(
         annotation_layer: dict[str, Any], force: bool
     ) -> dict[str, Any]:
-        chart = ChartDAO.find_by_id(annotation_layer["value"])
-        if not chart:
+        # pylint: disable=import-outside-toplevel,superfluous-parens
+        from superset.charts.data.commands.get_data_command import ChartDataCommand
+
+        if not (chart := ChartDAO.find_by_id(annotation_layer["value"])):
             raise QueryObjectValidationError(_("The chart does not exist"))
-        if not chart.datasource:
-            raise QueryObjectValidationError(_("The chart datasource does not exist"))
-        form_data = chart.form_data.copy()
-        form_data.update(annotation_layer.get("overrides", {}))
+
         try:
-            viz_obj = get_viz(
-                datasource_type=chart.datasource.type,
-                datasource_id=chart.datasource.id,
-                form_data=form_data,
-                force=force,
-            )
-            payload = viz_obj.get_payload()
-            return payload["data"]
+            if chart.viz_type in viz_types:
+                if not chart.datasource:
+                    raise QueryObjectValidationError(
+                        _("The chart datasource does not exist"),
+                    )
+
+                form_data = chart.form_data.copy()
+                form_data.update(annotation_layer.get("overrides", {}))
+
+                payload = get_viz(
+                    datasource_type=chart.datasource.type,
+                    datasource_id=chart.datasource.id,
+                    form_data=form_data,
+                    force=force,
+                ).get_payload()
+
+                return payload["data"]
+
+            if not (query_context := chart.get_query_context()):
+                raise QueryObjectValidationError(
+                    _("The chart query context does not exist"),
+                )
+
+            if overrides := annotation_layer.get("overrides"):
+                if time_grain_sqla := overrides.get("time_grain_sqla"):
+                    for query_object in query_context.queries:
+                        query_object.extras["time_grain_sqla"] = time_grain_sqla
+
+                if time_range := overrides.get("time_range"):
+                    from_dttm, to_dttm = get_since_until_from_time_range(time_range)
```

Comment:
The `QueryObject` has both `time_range` and (`from_dttm`, `to_dttm`) but I was only able to get the annotation override to adhere to the later.